crates/polars-plan/src/dsl/function_expr/temporal.rs

```diff
@@ -294,28 +319,87 @@ pub(super) fn temporal_range_dispatch(
     let start = start.get(0).unwrap().extract::<i64>().unwrap();
     let end = end.get(0).unwrap().extract::<i64>().unwrap();
 
-    let out = match dtype {
-        DataType::Date => date_range_impl(
-            name,
-            start,
-            end,
-            every,
-            closed,
-            TimeUnit::Milliseconds,
-            None,
-        )?,
+    let result = match dtype {
         DataType::Datetime(tu, ref tz) => {
-            date_range_impl(name, start, end, every, closed, tu, tz.as_ref())?
+            date_range_impl("date", start, end, interval, closed, tu, tz.as_ref())?
         },
         _ => unimplemented!(),
     };
-    Ok(out.cast(&dtype).unwrap().into_series())
+    Ok(result.cast(&dtype).unwrap().into_series())
 }
 
-pub(super) fn temporal_ranges_dispatch(
+pub(super) fn temporal_ranges(
     s: &[Series],
-    name: &str,
-    every: Duration,
+    interval: Duration,
+    closed: ClosedWindow,
+    time_unit: Option<TimeUnit>,
+    time_zone: Option<TimeZone>,
+) -> PolarsResult<Series> {
+    if s[0].dtype() == &DataType::Date && interval.nanoseconds() == 0 {
+        date_ranges(s, interval, closed)
+    } else {
+        datetime_ranges(s, interval, closed, time_unit, time_zone)
+    }
+}
+
+pub(super) fn date_ranges(
+    s: &[Series],
+    interval: Duration,
+    closed: ClosedWindow,
+) -> PolarsResult<Series> {
+    let start = &s[0];
+    let end = &s[1];
+
+    polars_ensure!(
+        start.len() == end.len(),
+        ComputeError: "`start` and `end` must have the same length",
+    );
+
+    let start = date_series_to_i64_ca(start)? * DAYS_TO_MILLISECONDS;
+    let end = date_series_to_i64_ca(end)? * DAYS_TO_MILLISECONDS;
+
+    let mut builder = ListPrimitiveChunkedBuilder::<Int32Type>::new(
+        "date_range",
+        start.len(),
+        start.len() * CAPACITY_FACTOR,
+        DataType::Int32,
+    );
+    for (start, end) in start.as_ref().into_iter().zip(&end) {
+        match (start, end) {
+            (Some(start), Some(end)) => {
+                let rng = date_range_impl(
+                    "",
+                    start,
+                    end,
+                    interval,
+                    closed,
+                    TimeUnit::Milliseconds,
+                    None,
+                )?;
+                let rng = rng.cast(&DataType::Date).unwrap();
```

Comment:
I really don't like this cast in the hot loop. I think we should have a `date_range` that returns `i32` or one that is generic over the integer type.

Reply:
Yeah, this is very wasteful. I was planning to address that in a separate PR by adding a date range that constructs a `i32` array directly.